modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryReducer.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.query;
+
+import java.util.Collection;
+import java.util.UUID;
+import org.apache.ignite.IgniteCheckedException;
+
+/**
+ * This class is responsible for reducing results of cache query. Query results are delivered via callback
+ * {@link #onPage(UUID, Collection, boolean)}.
+ *
+ * <T> is a type of cache query result item.
+ */
+public interface CacheQueryReducer<T> {
+    /**
+     * @return Next item.
+     */
+    public T next() throws IgniteCheckedException;
+
+    /**
+     * @return {@code true} if there is a next item, otherwise {@code false}.
+     */
+    public boolean hasNext() throws IgniteCheckedException;
+
+    /**
+     * Callback that invoked on receiving a new page.
+     *
+     * @param nodeId Node ID that sent this page.
+     * @param data Page data rows.
+     * @param last Whether this page is last for specified {@code nodeId}.
+     * @return {@code true} if this page is final page for query and no more pages are waited, otherwise {@code false}.
+     */
+    public boolean onPage(UUID nodeId, Collection<T> data, boolean last);
+
+    /**
+     * Callback in case of page with error.
+     */
+    public void onError();
+
+    /**
+     * Callback that invokes after a query future is done.
+     */
+    public void onFinish();
```

Comment:
Why do we need this method if we already have onError and onPage with 'last' parameter?

Reply:
There is a comment about it, see invokation [1]. We should release latch after onDone() invokation. Also I've described this in previous conversation: https://github.com/apache/ignite/pull/9081#discussion_r649824494

[1] [GridCacheQueryFutureAdapter:332](https://github.com/apache/ignite/blob/0add93b488ebe7265b2a0f951fcdfd55adbcc01f/modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java#L332)

Reply:
Is it possible to release the latch in onError() and onPage() method?
This will reduce `future` and `reducer` coupling.

Reply:
It's not possible, as it's required to release latch after `onDone()`. Actually there is a tough coupling of future and reducer. Don't think this is a problem

Reply:
We can move `onFinish()` from `onDone()` and explicity invoke it, but there are about 5-6 places to invoke this method - after every `onDone()`. Looks wrong to use this method such way